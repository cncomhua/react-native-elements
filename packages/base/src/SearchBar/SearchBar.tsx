import React from 'react';
import { SearchBarIOS, SearchBarIosProps } from './SearchBar-ios';
import { SearchBarAndroid, SearchBarAndroidProps } from './SearchBar-android';
import { SearchBarDefault, SearchBarDefaultProps } from './SearchBar-default';
import {
  ActivityIndicatorProps,
  StyleProp,
  TextStyle,
  ViewStyle,
  TextInput,
} from 'react-native';
import { IconNode } from '../Icon';
import { Theme } from '../helpers';

const SEARCH_BAR_COMPONENTS = {
  ios: SearchBarIOS,
  android: SearchBarAndroid,
  default: SearchBarDefault,
};

export interface SearchBarBaseProps
  extends React.ComponentPropsWithRef<typeof TextInput> {
  containerStyle?: StyleProp<ViewStyle>;
  inputContainerStyle?: StyleProp<ViewStyle>;
  clearIcon?: IconNode;
  searchIcon?: IconNode;
  inputStyle?: StyleProp<TextStyle>;
  loadingProps?: ActivityIndicatorProps;
  showLoading?: boolean;
  leftIconContainerStyle?: StyleProp<ViewStyle>;
  rightIconContainerStyle?: StyleProp<ViewStyle>;
  onClear?(): void;
  onFocus?(): void;
  onBlur?(): void;
  onChangeText?(text: string): void;
  onCancel?(): void;
  theme?: Theme;
}

export type SearchBarProps =
  | SearchBarDefaultProps
  | SearchBarAndroidProps
  | SearchBarIosProps;

export class SearchBar extends React.Component<
  SearchBarProps & {
    theme?: Theme;
    platform?: 'default' | 'ios' | 'android';
  }
> {
  searchBar!: SearchBarIOS;
  static defaultProps = {
    platform: 'default' as const,
  };

  focus = () => {
    this.searchBar.focus();
  };

  blur = () => {
    this.searchBar.blur();
  };

  clear = () => {
    this.searchBar.clear();
  };

  cancel = () => {
    this.searchBar?.cancel();
  };

  render() {
    const Component: typeof React.Component =
      SEARCH_BAR_COMPONENTS[this.props.platform] || SearchBarDefault;

    return (
      <Component
        ref={(ref: SearchBarIOS) => {
          this.searchBar = ref;
        }}
        {...this.props}
      />
    );
  }
}
