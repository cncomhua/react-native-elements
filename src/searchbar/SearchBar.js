import React from 'react';
import PropTypes from 'prop-types';

import IOSSearchBar from './SearchBar-ios';
import AndroidSearchBar from './SearchBar-android';
import DefaultSearchBar from './SearchBar-default';
import { merge, ThemeConsumer } from '../config';

const SEARCHBAR_COMPONENTS = {
  ios: IOSSearchBar,
  android: AndroidSearchBar,
  default: DefaultSearchBar,
};

const SearchBar = props => {
  const SearchBar = SEARCHBAR_COMPONENTS[props.platform] || DefaultSearchBar;
  return <SearchBar ref={ref => (props.searchBarRef = ref)} {...props} />;
};

SearchBar.propTypes = {
  searchBarRef: PropTypes.func,
  platform: PropTypes.oneOf(['default', 'ios', 'android']),
};

SearchBar.defaultProps = {
  platform: 'default',
};

export default class ThemedSearchBar extends React.Component {
  focus = () => {
    this.searchbar.focus();
  };

  blur = () => {
    this.searchbar.blur();
  };

  clear = () => {
    this.searchbar.clear();
  };

  cancel = () => {
    this.searchbar.cancel && this.searchbar.cancel();
  };

  render() {
    return (
      <ThemeConsumer>
        {({ theme }) => (
          <SearchBar
            searchBarRef={ref => (this.searchbar = ref)}
            {...merge({}, theme.SearchBar, this.props)}
            theme={theme}
          />
        )}
      </ThemeConsumer>
    );
  }
}
