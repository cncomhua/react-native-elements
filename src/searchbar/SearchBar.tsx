import React from 'react';
import { withTheme } from '../config';
import IOSSearchBar, { SearchBarIosProps } from './SearchBar-ios';
import AndroidSearchBar, { SearchBarAndroidProps } from './SearchBar-android';
import DefaultSearchBar, { SearchBarDefaultProps } from './SearchBar-default';
import {
  ActivityIndicatorProps,
  StyleProp,
  TextStyle,
  ViewStyle,
} from 'react-native';
import { IconNode } from '../icons/Icon';
import { Theme, FullTheme } from '../config/theme';
type RecursivePartial<T> = { [P in keyof T]?: RecursivePartial<T[P]> };
const SEARCHBAR_COMPONENTS = {
  ios: IOSSearchBar,
  android: AndroidSearchBar,
  default: DefaultSearchBar,
};

export type SearchBarBaseProps = {
  platform: 'default' | 'ios' | 'android';
  containerStyle?: StyleProp<ViewStyle>;
  inputContainerStyle?: StyleProp<ViewStyle>;
  clearIcon?: IconNode;
  searchIcon?: IconNode;
  inputStyle?: StyleProp<TextStyle>;
  loadingProps?: ActivityIndicatorProps;
  showLoading?: boolean;
  leftIconContainerStyle?: StyleProp<ViewStyle>;
  rightIconContainerStyle?: StyleProp<ViewStyle>;
  onClear?(): void;
  onFocus?(): void;
  onBlur?(): void;
  onChangeText?(text: string): void;
  onCancel?(): void;
  theme?: Theme;
  updateTheme: (updates: RecursivePartial<FullTheme>) => void;
  replaceTheme: (updates: RecursivePartial<FullTheme>) => void;
};

export type SearchBarProps = SearchBarBaseProps &
  SearchBarDefaultProps &
  SearchBarAndroidProps &
  SearchBarIosProps;

class SearchBar extends React.Component<SearchBarBaseProps> {
  searchbar!: IOSSearchBar;
  static defaultProps = {
    platform: 'default' as const,
  };

  focus = () => {
    this.searchbar.focus();
  };

  blur = () => {
    this.searchbar.blur();
  };

  clear = () => {
    this.searchbar.clear();
  };

  cancel = () => {
    this.searchbar.cancel && this.searchbar.cancel();
  };

  render() {
    const Component: typeof React.Component =
      SEARCHBAR_COMPONENTS[this.props.platform] || DefaultSearchBar;

    return (
      <Component
        // @ts-ignore
        ref={(ref: IOSSearchBar) => {
          this.searchbar = ref;
        }}
        {...this.props}
      />
    );
  }
}

export { SearchBar };
export default withTheme(SearchBar, 'SearchBar');
