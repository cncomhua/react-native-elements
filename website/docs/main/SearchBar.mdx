---
id: searchbar
title: SearchBar
slug: /searchbar
---

import Usage from './usage/SearchBar/SearchBar.mdx';

import Props from './props/SearchBar.mdx';

SearchBars are used to search or filter items. Use a SearchBar when the number
of items directly impacts a user's ability to find one of them.

## Usage

<Usage />

## Props

<Props />
