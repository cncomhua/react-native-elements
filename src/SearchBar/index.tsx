import { withTheme } from '../config';
import { SearchBar, SearchBarProps } from './SearchBar';
import { SearchBarAndroidProps } from './SearchBar-android';
import { SearchBarIosProps } from './SearchBar-ios';
import { SearchBarDefaultProps } from './SearchBar-default';

export { SearchBar };
export type {
  SearchBarProps,
  SearchBarAndroidProps,
  SearchBarDefaultProps,
  SearchBarIosProps,
};
/**
 * RNE SearchBar
 *
 * **Preview**
 *
 * ![URL](URL)
 *
 * [**Documentation**](https://reactnativeelements.com/docs/searchbar)
 */
export default withTheme(SearchBar, 'SearchBar');
