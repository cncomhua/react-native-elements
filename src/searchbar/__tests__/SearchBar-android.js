import React from 'react';
import { shallow } from 'enzyme';
import toJson from 'enzyme-to-json';
import SearchBar from '../SearchBar-android';

describe('Android SearchBar component', () => {
  it('should render without issues', () => {
    const component = shallow(<SearchBar />);

    expect(component.length).toBe(1);
    expect(toJson(component)).toMatchSnapshot();
  });

  it('should render with loading', () => {
    const component = shallow(
      <SearchBar
        showLoading
        loadingProps={{
          style: { flex: 1 },
        }}
      />
    );

    expect(component.length).toBe(1);
    expect(toJson(component)).toMatchSnapshot();
  });

  it('should render with clear icon', () => {
    const component = shallow(<SearchBar value="test" />);

    expect(component.length).toBe(1);
    expect(toJson(component)).toMatchSnapshot();
  });

  it('should render with clear icon with defaultValue', () => {
    const component = shallow(<SearchBar defaultValue="test" />);

    expect(component.length).toBe(1);
    expect(toJson(component)).toMatchSnapshot();
  });

  it('should render without search icon', () => {
    const component = shallow(<SearchBar noIcon />);

    expect(component.length).toBe(1);
    expect(toJson(component)).toMatchSnapshot();
  });

  it('should render without clear icon', () => {
    const component = shallow(<SearchBar clearIcon={false} />);

    expect(component.length).toBe(1);
    expect(toJson(component)).toMatchSnapshot();
  });

  it('should render with input container style', () => {
    const component = shallow(
      <SearchBar
        inputContainerStyle={{
          backgroundColor: 'red',
        }}
      />
    );

    expect(component.length).toBe(1);
    expect(toJson(component)).toMatchSnapshot();
  });

  it('should render with container style', () => {
    const component = shallow(
      <SearchBar
        containerStyle={{
          paddingTop: 0,
          paddingBottom: 0,
          backgroundColor: 'gray',
        }}
      />
    );

    expect(component.length).toBe(1);
    expect(toJson(component)).toMatchSnapshot();
  });

  it('should call onFocus when input is focused', () => {
    const onFocusMock = jest.fn();
    const component = shallow(<SearchBar onFocus={onFocusMock} />);
    component.find('Input').simulate('focus');
    expect(onFocusMock).toBeCalled();
  });

  it('should call onBlur when input is blured', () => {
    const onBlurMock = jest.fn();
    const component = shallow(<SearchBar onFocus={onBlurMock} />);
    component.find('Input').simulate('focus');
    component.find('Input').simulate('blur');
    expect(onBlurMock).toBeCalled();
  });

  it('should call onChangeText when input is changed', () => {
    const onChangeMock = jest.fn();
    const component = shallow(<SearchBar onChangeText={onChangeMock} />);
    component.find('Input').simulate('changeText', 'test');
    expect(onChangeMock).toBeCalled();
  });
});
