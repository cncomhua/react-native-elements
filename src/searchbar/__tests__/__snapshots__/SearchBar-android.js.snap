// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Android SearchBar component should render with clear icon 1`] = `
<View
  style={
    Array [
      Object {
        "backgroundColor": "white",
        "paddingBottom": 8,
        "paddingTop": 8,
        "width": 750,
      },
      undefined,
    ]
  }
>
  <Input
    containerStyle={
      Array [
        Object {
          "borderBottomWidth": 0,
          "width": 750,
        },
        undefined,
      ]
    }
    inputStyle={
      Array [
        Object {
          "flex": 1,
          "marginLeft": 24,
          "marginRight": 8,
        },
        undefined,
      ]
    }
    leftIcon={
      <Icon
        allowFontScaling={false}
        color="rgba(0, 0, 0, 0.54)"
        name="magnify"
        onPress={null}
        size={25}
      />
    }
    leftIconContainerStyle={
      Array [
        Object {
          "marginLeft": 8,
        },
        undefined,
      ]
    }
    onBlur={[Function]}
    onCancel={[Function]}
    onChangeText={[Function]}
    onClearText={[Function]}
    onFocus={[Function]}
    rightIcon={
      <View
        style={
          Object {
            "flexDirection": "row",
          }
        }
      >
        <Icon
          allowFontScaling={false}
          color="rgba(0, 0, 0, 0.54)"
          name="close"
          onPress={[Function]}
          size={25}
        />
      </View>
    }
    rightIconContainerStyle={
      Array [
        Object {
          "marginRight": 8,
        },
        undefined,
      ]
    }
    value="test"
  />
</View>
`;

exports[`Android SearchBar component should render with clear icon with defaultValue 1`] = `
<View
  style={
    Array [
      Object {
        "backgroundColor": "white",
        "paddingBottom": 8,
        "paddingTop": 8,
        "width": 750,
      },
      undefined,
    ]
  }
>
  <Input
    containerStyle={
      Array [
        Object {
          "borderBottomWidth": 0,
          "width": 750,
        },
        undefined,
      ]
    }
    defaultValue="test"
    inputStyle={
      Array [
        Object {
          "flex": 1,
          "marginLeft": 24,
          "marginRight": 8,
        },
        undefined,
      ]
    }
    leftIcon={
      <Icon
        allowFontScaling={false}
        color="rgba(0, 0, 0, 0.54)"
        name="magnify"
        onPress={null}
        size={25}
      />
    }
    leftIconContainerStyle={
      Array [
        Object {
          "marginLeft": 8,
        },
        undefined,
      ]
    }
    onBlur={[Function]}
    onCancel={[Function]}
    onChangeText={[Function]}
    onClearText={[Function]}
    onFocus={[Function]}
    rightIcon={
      <View
        style={
          Object {
            "flexDirection": "row",
          }
        }
      >
        <Icon
          allowFontScaling={false}
          color="rgba(0, 0, 0, 0.54)"
          name="close"
          onPress={[Function]}
          size={25}
        />
      </View>
    }
    rightIconContainerStyle={
      Array [
        Object {
          "marginRight": 8,
        },
        undefined,
      ]
    }
  />
</View>
`;

exports[`Android SearchBar component should render with container style 1`] = `
<View
  style={
    Array [
      Object {
        "backgroundColor": "white",
        "paddingBottom": 8,
        "paddingTop": 8,
        "width": 750,
      },
      Object {
        "backgroundColor": "gray",
        "paddingBottom": 0,
        "paddingTop": 0,
      },
    ]
  }
>
  <Input
    containerStyle={
      Array [
        Object {
          "borderBottomWidth": 0,
          "width": 750,
        },
        undefined,
      ]
    }
    inputStyle={
      Array [
        Object {
          "flex": 1,
          "marginLeft": 24,
          "marginRight": 8,
        },
        undefined,
      ]
    }
    leftIcon={
      <Icon
        allowFontScaling={false}
        color="rgba(0, 0, 0, 0.54)"
        name="magnify"
        onPress={null}
        size={25}
      />
    }
    leftIconContainerStyle={
      Array [
        Object {
          "marginLeft": 8,
        },
        undefined,
      ]
    }
    onBlur={[Function]}
    onCancel={[Function]}
    onChangeText={[Function]}
    onClearText={[Function]}
    onFocus={[Function]}
    rightIcon={
      <View
        style={
          Object {
            "flexDirection": "row",
          }
        }
      />
    }
    rightIconContainerStyle={
      Array [
        Object {
          "marginRight": 8,
        },
        undefined,
      ]
    }
  />
</View>
`;

exports[`Android SearchBar component should render with input container style 1`] = `
<View
  style={
    Array [
      Object {
        "backgroundColor": "white",
        "paddingBottom": 8,
        "paddingTop": 8,
        "width": 750,
      },
      undefined,
    ]
  }
>
  <Input
    containerStyle={
      Array [
        Object {
          "borderBottomWidth": 0,
          "width": 750,
        },
        Object {
          "backgroundColor": "red",
        },
      ]
    }
    inputStyle={
      Array [
        Object {
          "flex": 1,
          "marginLeft": 24,
          "marginRight": 8,
        },
        undefined,
      ]
    }
    leftIcon={
      <Icon
        allowFontScaling={false}
        color="rgba(0, 0, 0, 0.54)"
        name="magnify"
        onPress={null}
        size={25}
      />
    }
    leftIconContainerStyle={
      Array [
        Object {
          "marginLeft": 8,
        },
        undefined,
      ]
    }
    onBlur={[Function]}
    onCancel={[Function]}
    onChangeText={[Function]}
    onClearText={[Function]}
    onFocus={[Function]}
    rightIcon={
      <View
        style={
          Object {
            "flexDirection": "row",
          }
        }
      />
    }
    rightIconContainerStyle={
      Array [
        Object {
          "marginRight": 8,
        },
        undefined,
      ]
    }
  />
</View>
`;

exports[`Android SearchBar component should render with loading 1`] = `
<View
  style={
    Array [
      Object {
        "backgroundColor": "white",
        "paddingBottom": 8,
        "paddingTop": 8,
        "width": 750,
      },
      undefined,
    ]
  }
>
  <Input
    containerStyle={
      Array [
        Object {
          "borderBottomWidth": 0,
          "width": 750,
        },
        undefined,
      ]
    }
    inputStyle={
      Array [
        Object {
          "flex": 1,
          "marginLeft": 24,
          "marginRight": 8,
        },
        undefined,
      ]
    }
    leftIcon={
      <Icon
        allowFontScaling={false}
        color="rgba(0, 0, 0, 0.54)"
        name="magnify"
        onPress={null}
        size={25}
      />
    }
    leftIconContainerStyle={
      Array [
        Object {
          "marginLeft": 8,
        },
        undefined,
      ]
    }
    onBlur={[Function]}
    onCancel={[Function]}
    onChangeText={[Function]}
    onClearText={[Function]}
    onFocus={[Function]}
    rightIcon={
      <View
        style={
          Object {
            "flexDirection": "row",
          }
        }
      >
        <ActivityIndicator
          animating={true}
          color="#999999"
          hidesWhenStopped={true}
          size="small"
          style={
            Array [
              false,
              Object {
                "flex": 1,
              },
            ]
          }
        />
      </View>
    }
    rightIconContainerStyle={
      Array [
        Object {
          "marginRight": 8,
        },
        undefined,
      ]
    }
  />
</View>
`;

exports[`Android SearchBar component should render without clear icon 1`] = `
<View
  style={
    Array [
      Object {
        "backgroundColor": "white",
        "paddingBottom": 8,
        "paddingTop": 8,
        "width": 750,
      },
      undefined,
    ]
  }
>
  <Input
    containerStyle={
      Array [
        Object {
          "borderBottomWidth": 0,
          "width": 750,
        },
        undefined,
      ]
    }
    inputStyle={
      Array [
        Object {
          "flex": 1,
          "marginLeft": 24,
          "marginRight": 8,
        },
        undefined,
      ]
    }
    leftIcon={
      <Icon
        allowFontScaling={false}
        color="rgba(0, 0, 0, 0.54)"
        name="magnify"
        onPress={null}
        size={25}
      />
    }
    leftIconContainerStyle={
      Array [
        Object {
          "marginLeft": 8,
        },
        undefined,
      ]
    }
    onBlur={[Function]}
    onCancel={[Function]}
    onChangeText={[Function]}
    onClearText={[Function]}
    onFocus={[Function]}
    rightIcon={
      <View
        style={
          Object {
            "flexDirection": "row",
          }
        }
      />
    }
    rightIconContainerStyle={
      Array [
        Object {
          "marginRight": 8,
        },
        undefined,
      ]
    }
  />
</View>
`;

exports[`Android SearchBar component should render without issues 1`] = `
<View
  style={
    Array [
      Object {
        "backgroundColor": "white",
        "paddingBottom": 8,
        "paddingTop": 8,
        "width": 750,
      },
      undefined,
    ]
  }
>
  <Input
    containerStyle={
      Array [
        Object {
          "borderBottomWidth": 0,
          "width": 750,
        },
        undefined,
      ]
    }
    inputStyle={
      Array [
        Object {
          "flex": 1,
          "marginLeft": 24,
          "marginRight": 8,
        },
        undefined,
      ]
    }
    leftIcon={
      <Icon
        allowFontScaling={false}
        color="rgba(0, 0, 0, 0.54)"
        name="magnify"
        onPress={null}
        size={25}
      />
    }
    leftIconContainerStyle={
      Array [
        Object {
          "marginLeft": 8,
        },
        undefined,
      ]
    }
    onBlur={[Function]}
    onCancel={[Function]}
    onChangeText={[Function]}
    onClearText={[Function]}
    onFocus={[Function]}
    rightIcon={
      <View
        style={
          Object {
            "flexDirection": "row",
          }
        }
      />
    }
    rightIconContainerStyle={
      Array [
        Object {
          "marginRight": 8,
        },
        undefined,
      ]
    }
  />
</View>
`;

exports[`Android SearchBar component should render without search icon 1`] = `
<View
  style={
    Array [
      Object {
        "backgroundColor": "white",
        "paddingBottom": 8,
        "paddingTop": 8,
        "width": 750,
      },
      undefined,
    ]
  }
>
  <Input
    containerStyle={
      Array [
        Object {
          "borderBottomWidth": 0,
          "width": 750,
        },
        undefined,
      ]
    }
    inputStyle={
      Array [
        Object {
          "flex": 1,
          "marginLeft": 24,
          "marginRight": 8,
        },
        undefined,
      ]
    }
    leftIconContainerStyle={
      Array [
        Object {
          "marginLeft": 8,
        },
        undefined,
      ]
    }
    onBlur={[Function]}
    onCancel={[Function]}
    onChangeText={[Function]}
    onClearText={[Function]}
    onFocus={[Function]}
    rightIcon={
      <View
        style={
          Object {
            "flexDirection": "row",
          }
        }
      />
    }
    rightIconContainerStyle={
      Array [
        Object {
          "marginRight": 8,
        },
        undefined,
      ]
    }
  />
</View>
`;
